### ISSUE TITLE
Different behavior with custom loss function.

### ISSUE THREAD

**Issue by EhsanEI** [fchollet/keras#4108]
Hi,

I wrote a custom loss function that just uses the built-in binary crossentropy loss:

``` python
from keras.objectives import get as get_objective
def my_loss(y_true, y_pred):
    co = get_objective('binary_crossentropy')
    return co(y_true, y_pred)
```

However, the behavior is different when I use this function instead of passing 'binary_crossentropy' to the model's compile method. This function yields worse results. I suspect it might be a numerical problem but I don't know how to fix it.

In [this gist,](https://gist.github.com/EhsanEI/90203acd0026d915699a7cbb9af584a9) I get ~99% validation accuracy on MNIST with the built-in function and ~96% validation accuracy with my_loss after the first epoch. The difference is much bigger on my own dataset.

**Comment by JimBiardCics** [comment]
I have seen pretty much the same problem. I wrote this function:
```python
def mcc(output, target, from_logits=False):
    if from_logits:
        output = T.nnet.softmax(output)
    else:
        # scale preds so that the class probas of each sample sum to 1
        output /= output.sum(axis=-1, keepdims=True)

    # avoid numerical instability with _EPSILON clipping
    output = T.clip(output, keras.backend.common._EPSILON, 1.0 - keras.backend.common._EPSILON)
    
    return T.nnet.categorical_crossentropy(output, target)
```
which is a near-exact copy of the function from the keras source, but when I compile with loss=mcc instead of loss='categorical_crossentropy' the loss drops to ~1e-14 within two epochs.

I'm using keras 1.1.0 in an ipython notebook.

**Comment by NakramR** [comment]
I'm seeing the exact same issue, also with binary_crossentropy.

I copied the code for the function from https://github.com/fchollet/keras/blob/master/keras/losses.py , added "my_" in front 

def my_binary_crossentropy(y_true, y_pred):
    return K.mean(K.binary_crossentropy(y_true, y_pred), axis=-1)

But calling my version of the cross-entropy instead of passing the string "binary_crossentropy" yields vastly different results.

autoencoder.compile(optimizer='adadelta', loss='binary_crossentropy') # works
autoencoder.compile(optimizer='adadelta', loss=my_binary_crossentropy) # sucks

**Comment by m30m** [comment]
I'm also seeing the same issue. I guess this is related to #6680

**Comment by mha-py** [comment]
I had exactly the same issue, metrics showing the same value but accuracy in the custom case sucks. It looks like the training is working equally, but the accuracy metric is not working. I could fix it by using

`def acc(y_true, y_pred):
    return K.mean(K.round(y_pred)*y_true + (1.-K.round(y_pred))*(1.-y_true))`

as the accuracy (which gives exactly the same as metric=['accuracy'] in the normal case but is not buggy with the custom loss).

**Comment by CA4GitHub** [comment]
@NakramR did you ever find a solution?

**Comment by JimBiardCics** [comment]
@NakramR @CA4GitHub One possible reason for the problem in the example given above by @NakramR is that the cross-entropy function does not return a mean. It returns a tensor that is averaged later by higher level code to produce a single loss value. I'm not sure, but I believe this is related to using the output from the function for back-propagation.

**Comment by EhsanEI** [comment]
@CA4GitHub Yes. I was new to Keras back then and I had mistakenly used this "binary-crossentropy" loss for a multi-class classification task. IIRC the accuracy metric when paired with binary-crossentropy loss assumed two classes or a different kind of encoding in the labels, so the network was trained well but the reported accuracy was wrong. I might be misremembering some details but this is the overall idea.

**Comment by sreenivasaupadhyaya** [comment]
Guys,

Did anyoe have a convincing solution to make custom_binarycrossentropy work?

I tried all possible methods (even making the whole training data size same as the bacth size to eliminate the dependence on global averaging during batch wise processing.). But i see significant difference between my binary cross entropy implementation and the one from keras ( by specifying loss = 'binary_crossentropy')

My crustom binary cross entropy code is as follows
def _loss_tensor(y_true, y_pred):
    y_pred = K.clip(y_pred, _EPSILON, 1.0-_EPSILON)
    out = (y_true * K.log(y_pred) + (1.0 - y_true) * K.log(1.0 - y_pred))
    return -K.mean(out)
def _loss_tensor2(y_true, y_pred):
    y_pred = K.clip(y_pred, _EPSILON, 1.0-_EPSILON)
    out = -(y_true * K.log(y_pred) + -(1.0 - y_true) * K.log(1.0 - y_pred))
    return out
def _loss_tensor2(y_true, y_pred):
    loss1 = K.binary_crossentropy(y_true, y_pred)
    return loss1
None of these methods work. It doesnt work even if i do K.mean() before ir eturn the results from custom loss function. 

I am not able to understand what special does using loss = 'binary_crossentropy' does. When i use my custom loss function , the training sucks and it does work as expected.

I need my custom loss function to manipulate the loss function depending on the error and penalizing a certain type of classification error more.

**Comment by JimBiardCics** [comment]
@sreenivasaupadhyaya I think your _loss_tensor2 function should work, but there is a mistake in the equation. It should be

out = -(y_true * K.log(y_pred) + (1.0 - y_true) * K.log(1.0 - y_pred))

Another thing to consider is that y_pred must have passed through a softmax activation layer or otherwise have values that take the form of probabilities before the binary cross-entropy is calculated.

Your first function will not work correctly because it takes a mean. It seems that the third function should work, so it makes me wonder what form your y_pred tensor has.

**Comment by sreenivasaupadhyaya** [comment]
@JimBiardCics  Thanks a lot for your time to solve the issue.
Yes, my model is a binary classifier with Dense layer of size 1 with sigmoid activation. Hence the custom loss function input y_pred will be probability values which is necessary for computing the  bianry cross entropy.

Secondly, The formaula you correction you mentioned (adding a negative sign), is already taken care in my formula but its rather applied separately to the two parts of the calculation like

" **-**(y_true * K.log(y_pred) + **-**(1.0 - y_true) * K.log(1.0 - y_pred)) "

Do you have any comments for the implementation below:
`def _loss_tensor2(y_true, y_pred):
loss1 = K.binary_crossentropy(y_true, y_pred)
return loss1`

Looking forward to your response.

**Comment by JimBiardCics** [comment]
@sreenivasaupadhyaya It may be correct in your code, but the equation in your second function is not correct. What you wrote is
> out = -(y_true * K.log(y_pred) + -(1.0 - y_true) * K.log(1.0 - y_pred))

Breaking out with indentation, you can see

out = -(
        y_true * K.log(y_pred) +
        -(1.0 - y_true) * K.log(1.0 - y_pred)
)

There is an excess minus sign on the second term.

Regarding the _loss_tensor2 function above, I would tend to think that it should work. If it doesn't then there is probably something unexpected in your code. If you'd like to share your code, I'd be willing to take a look at it.

**Comment by sreenivasaupadhyaya** [comment]
@JimBiardCics  Thanks again.
you were right , indeed i misplaced the - sign when i put the code here.

Here is the test setup i was using,

```
import tensorflow as tf
import keras
import keras.backend as K
y_true = tf.convert_to_tensor(np.array([1,1,1,1,1,0,0,0,0,0,1,1,1,1,1,0,0,0,0,0]))
y_true = K.cast(y_true, dtype ='float32')
y_pred1   = tf.convert_to_tensor(np.array([0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.50,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0]))
y_pred1 = K.cast(y_pred1, dtype ='float32')
loss_reweight = reweight2(y_true,y_pred)
loss1 = K.binary_crossentropy(y_true,y_pred)
loss2 = _loss_tensor2(y_true,y_pred)
print('loss reweight is ', loss_reweight)
print('loss1 is ',loss1)
print('loss2 is ',loss2)



_EPSILON = K.epsilon()#1e-7
def _loss_tensor2(y_true, y_pred):
    y_pred = K.clip(y_pred, _EPSILON, 1.0-_EPSILON)
    out = -(y_true * K.log(y_pred) + (1.0 - y_true) * K.log(1.0 - y_pred))
    return out

```

I couldnt get the same result as K.binary_crossentropy() using the custom code. The end goal of this exercise was to write a custom loss function to weight (TP,tn,FP and FN) differently.
finally I got the custom code working and i used the inbuilt K.binary_crossentropy in the custom function. I am sharing the function below

`def reweight2(y_true, y_pred, tp_weight=1.0, tn_weight=1.0, fp_weight=1.0, fn_weight=1.0):
    # Get predictions
    y_true = K.cast(y_true,dtype='float32')
    y_pred = K.cast(y_pred, dtype='float32')

    y_pred_classes = K.greater_equal(y_pred, 0.5)
    y_pred_classes_float = K.cast(y_pred_classes, dtype ='float32')


    # Get misclassified examples
    wrongly_classified = K.not_equal(y_true, y_pred_classes_float)
    wrongly_classified_float = K.cast(wrongly_classified, dtype ='float32')

    # Get correctly classified examples
    correctly_classified = K.equal(y_true, y_pred_classes_float)
    correctly_classified_float = K.cast(correctly_classified, dtype ='float32')

    loss = K.binary_crossentropy(y_true,y_pred)
    # Get tp, fp, tn, fn
    tp = tf.math.multiply( y_true , loss )#tf.math.multiply( y_true , K.log(y_pred))
    tp = tf.math.multiply(tp,correctly_classified_float)

    tn = tf.math.multiply(loss, (1-y_true))#tf.math.multiply((1 - y_true) , K.log(1-y_pred))
    tn = tf.math.multiply(tn,correctly_classified_float)

    fn = tf.math.multiply(loss,y_true)# tf.math.multiply(y_true , K.log(y_pred))
    fn= tf.math.multiply(fn,wrongly_classified_float)


    fp =  tf.math.multiply(loss,(1-y_true))#tf.math.multiply((1 - y_true) , K.log(1-y_pred))
    fp = tf.math.multiply(fp,wrongly_classified_float)

    # Get weights
    weight_tensor =  tf.math.add_n([tf.math.multiply(tp_weight , tp) , tf.math.multiply(fp_weight , fp) , tf.math.multiply(tn_weight , tn) , tf.math.multiply(fn_weight , fn)])
    return weight_tensor`

**some of the type coversions and some lines are redundant/unnecessary. the earlier attempts were to make the function working with formula method and not use inbuilt K.binary_crossentropy!!

@JimBiardCics  let me know your comments.

**Comment by JimBiardCics** [comment]
@sreenivasaupadhyaya Part of the problem may be a mismatch between tensorflow and keras depending on the versions used. There are internal changes (I think) in tensorflow 2.0+ that could cause problems if you try to use independent keras with tensorflow. You should be sure to use the tensorflow version of keras. The other thing to keep in mind is that the result of a function composed of tensorflow objects and operations is a tensor has to be evaluated before you can tell what's going on. The tensorflow.keras.backend.eval() function produces a numpy array from a tensor result. I slightly modified your test code and found that the results from all loss functions were nearly identical. The only difference was for the elements where both y_true and y_pred were zero. There is  a 1.9e-8 difference between those elements of loss1 and loss2. This is due to the slightly different way the binary cross-entropy function is written in tensorflow. You can see the difference at [this link](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/backend.py#L4728).

I added a loss_tensor3 function that calculates the binary cross-entropy the same way tensorflow does. Here's the code that worked for me on my Mac.
```python
import numpy as np
import tensorflow as tf
from tensorflow import keras
import tensorflow.keras.backend as K

_EPSILON = K.epsilon()#1e-7


def reweight2(y_true, y_pred, tp_weight=1.0, tn_weight=1.0, fp_weight=1.0, fn_weight=1.0):
    # Get predictions
    y_pred_classes = K.greater_equal(y_pred, 0.5)
    y_pred_classes_float = K.cast(y_pred_classes, dtype ='float32')

    # Get misclassified examples
    wrongly_classified = K.not_equal(y_true, y_pred_classes_float)
    wrongly_classified_float = K.cast(wrongly_classified, dtype ='float32')

    # Get correctly classified examples
    correctly_classified = K.equal(y_true, y_pred_classes_float)
    correctly_classified_float = K.cast(correctly_classified, dtype ='float32')

    loss = K.binary_crossentropy(y_true,y_pred)

    # Get tp, fp, tn, fn
    tp = tf.math.multiply( y_true , loss )#tf.math.multiply( y_true , K.log(y_pred))
    tp = tf.math.multiply(tp,correctly_classified_float)

    tn = tf.math.multiply(loss, (1-y_true))#tf.math.multiply((1 - y_true) , K.log(1-y_pred))
    tn = tf.math.multiply(tn,correctly_classified_float)

    fn = tf.math.multiply(loss,y_true)# tf.math.multiply(y_true , K.log(y_pred))
    fn= tf.math.multiply(fn,wrongly_classified_float)


    fp =  tf.math.multiply(loss,(1-y_true))#tf.math.multiply((1 - y_true) , K.log(1-y_pred))
    fp = tf.math.multiply(fp,wrongly_classified_float)

    # Get weights
    weight_tensor =  tf.math.add_n([tf.math.multiply(tp_weight , tp) , tf.math.multiply(fp_weight , fp) , tf.math.multiply(tn_weight , tn) , tf.math.multiply(fn_weight , fn)])
    return weight_tensor


def _loss_tensor2(y_true, y_pred):
    y_pred = K.clip(y_pred, _EPSILON, 1.0 - _EPSILON)
    out    = -(y_true * K.log(y_pred) + (1.0 - y_true) * K.log(1.0 - y_pred))

    return out


def _loss_tensor3(y_true, y_pred):
    y_pred = K.clip(y_pred, _EPSILON, 1.0 - _EPSILON)
    out    = -(y_true * K.log(y_pred + _EPSILON) + (1.0 - y_true) * K.log(1.0 - y_pred + _EPSILON))

    return out


y_true = tf.convert_to_tensor(np.array([1,1,1,1,1,0,0,0,0,0,1,1,1,1,1,0,0,0,0,0]))
y_true = K.cast(y_true, dtype =np.float32)

y_pred = tf.convert_to_tensor(np.array([0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.5,0.50,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0]))
y_pred = K.cast(y_pred, dtype =np.float32)

loss_reweight = K.eval(reweight2(y_true,y_pred))

loss1 = K.eval(K.binary_crossentropy(y_true,y_pred))
loss2 = K.eval(_loss_tensor2(y_true,y_pred))
loss3 = K.eval(_loss_tensor3(y_true,y_pred))

print('loss reweight is ', loss_reweight)
print('\nloss1 is ',loss1)
print('\nloss2 is ',loss2)
print('\nloss3 is ',loss3)
print('\ndiff loss_reweight - loss1 is ', (loss_reweight - loss1))
print('\ndiff loss1 - loss2 is ', (loss1 - loss2))
print('\ndiff loss1 - loss3 is ', (loss1 - loss3))
print('\ndiff loss2 - loss3 is ', (loss2 - loss3))
```

**Comment by sreenivasaupadhyaya** [comment]
@JimBiardCics Thanks for trying out. I just repeated your experiment and found similar results.
However, when 'K.binary_crossentropy' or '_loss_tensor3' used as loss function in Keras model yields a different result.

Moreover, '_loss_tensor3' when used in loss and metrics filed in Keras model compilation gives different values which were quite surprising.

However, I achieved what I wanted using the 'reweight2' function. Like you suggested, there might be possible mismatch in keras and tf versions which in turn results in a different way of handling these cases.

**Comment by AvadSharma** [comment]
@JimBiardCics @sreenivasaupadhyaya : I am sorry I am new to ML but did you try something like following? 

def custom_loss(y_actual, y_predicted):
    return K.mean(K.binary_crossentropy(y_actual, y_predicted), axis=-1)

def acc(y_true, y_pred): return K.mean(K.round(y_pred)*y_true + (1.-K.round(y_pred))*(1.-y_true))

model.compile(loss=custom_loss,optimizer=opt,metrics=[acc])

I think you need to change your metrics because Keras automatically selects which accuracy implementation to use according to the loss, and this won't work if you use a custom loss.
